aws/resource_aws_sagemaker_notebook_instance.go

```diff
@@ -0,0 +1,371 @@
+		}
+
+		if _, err := conn.UpdateNotebookInstance(updateOpts); err != nil {
+			return fmt.Errorf("Error updating Sagemaker Notebook Instance: %s", err)
+		}
+
+		// Restart if needed
+		if previousStatus == "InService" {
+			startOpts := &sagemaker.StartNotebookInstanceInput{
+				NotebookInstanceName: aws.String(d.Id()),
+			}
+
+			if _, err := conn.StartNotebookInstance(startOpts); err != nil {
+				log.Printf("[DEBUG] Sagemaker Notebook Instance (%s) unable to restart", d.Id())
+			} else if err := waitSagemakerNotebookInstanceStatus(conn, d, "InService", "Failed"); err != nil {
+				log.Printf("[ERR] Sagemaker Notebook Instance (%s) did not start", d.Id())
+			}
+		}
+	}
+
+	d.Partial(false)
+
+	return resourceAwsSagemakerNotebookInstanceRead(d, meta)
+}
+
+func resourceAwsSagemakerNotebookInstanceDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+
+	if err := stopSagemakerNotebookInstance(conn, d, true); err != nil {
+		return err
+	}
+
+	deleteOpts := &sagemaker.DeleteNotebookInstanceInput{
+		NotebookInstanceName: aws.String(d.Id()),
+	}
+
+	if _, err := conn.DeleteNotebookInstance(deleteOpts); err != nil {
+		return err
+	}
+
+	return resource.Retry(10*time.Minute, func() *resource.RetryError {
+		_, status, _ := SagemakerNotebookInstanceStateRefreshFunc(conn, d.Id())()
+
+		if status == "" {
+			log.Printf("[DEBUG] Sagemaker Notebook Instance (%s) deleted", d.Id())
+			return nil
+		}
+
+		return resource.RetryableError(fmt.Errorf("[DEBUG] Waiting for Sagemaker Notebook Instance (%s) to be deleted", d.Id()))
+	})
+}
+
+func SagemakerNotebookInstanceStateRefreshFunc(conn *sagemaker.SageMaker, name string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		describeNotebookInput := &sagemaker.DescribeNotebookInstanceInput{
+			NotebookInstanceName: aws.String(name),
+		}
+		notebook, err := conn.DescribeNotebookInstance(describeNotebookInput)
+		if err != nil {
+			if sagemakerErr, ok := err.(awserr.Error); ok && sagemakerErr.Code() == "ResourceNotFound" {
+				notebook = nil
+			} else {
+				log.Printf("Error on SagemakerNotebookInstanceStateRefreshFunc: %s", err)
+				return nil, "", err
+			}
+		}
+
+		if notebook == nil {
+			return nil, "", nil
+		}
+
+		return notebook, *notebook.NotebookInstanceStatus, nil
+	}
+}
+
+func stopSagemakerNotebookInstance(conn *sagemaker.SageMaker, d *schema.ResourceData, wait bool) error {
+	stopOpts := &sagemaker.StopNotebookInstanceInput{
+		NotebookInstanceName: aws.String(d.Id()),
+	}
+
+	if _, err := conn.StopNotebookInstance(stopOpts); err != nil {
+		return fmt.Errorf("Error stopping Sagemaker Notebook Instance: %s", err)
+	}
+
+	if !wait {
+		return nil
+	}
+
+	return waitSagemakerNotebookInstanceStatus(conn, d, "Stopped")
+}
+
+func waitSagemakerNotebookInstanceStatus(conn *sagemaker.SageMaker, d *schema.ResourceData, desiredStatus ...string) error {
```

Comment:
As noted above we should swap `d` for id since that is all `d` is being used for here